tlsn/tlsn-notary-client/src/client.rs

```diff
@@ -207,193 +215,167 @@ impl NotaryClient {
+                .header("Upgrade", "TCP")
+                .body(Either::Right(Empty::<Bytes>::new()))
+                .map_err(|err| {
+                    error!("Failed to build http request for notarization");
+                    ClientError::new(ErrorKind::Internal, Some(Box::new(err)))
+                })?;
 
-        #[cfg(feature = "tracing")]
-        debug!("Sent notarization request");
-
-        if notarization_response.status() != StatusCode::SWITCHING_PROTOCOLS {
-            return Err(ClientError::new(
-                ErrorKind::NotarizationRequest,
-                Some(format!(
-                    "Notarization response is not SWITCHING_PROTOCOL: {:?}",
-                    notarization_response
-                )),
-                None,
-            ));
-        }
+            #[cfg(feature = "tracing")]
+            debug!("Sending notarization request: {:?}", notarization_request);
 
-        // Claim back notary socket after HTTP exchange is done
-        let Parts {
-            io: notary_socket, ..
-        } = notary_connection_task
-            .await
-            .map_err(|err| {
-                ClientError::new(
-                    ErrorKind::Unexpected,
-                    Some("Error when joining notary connection task".to_string()),
-                    Some(Box::new(err)),
-                )
-            })?
-            .map_err(|err| {
-                ClientError::new(
-                    ErrorKind::Unexpected,
+            let notarization_response = notary_request_sender
+                .send_request(notarization_request)
+                .await
+                .map_err(|err| {
+                    error!("Failed to send http request for notarization");
+                    ClientError::new(ErrorKind::Http, Some(Box::new(err)))
+                })?;
+
+            #[cfg(feature = "tracing")]
+            debug!("Sent notarization request");
+
+            if notarization_response.status() != StatusCode::SWITCHING_PROTOCOLS {
+                return Err(ClientError::new(
+                    ErrorKind::Internal,
                     Some(
-                        "Failed to claim back notary socket after HTTP exchange is done"
-                            .to_string(),
+                        format!(
+                            "Notarization response status is not SWITCHING_PROTOCOL: {:?}",
+                            notarization_response
+                        )
+                        .into(),
                     ),
-                    Some(Box::new(err)),
-                )
-            })?;
+                ));
+            }
+
+            Ok(configuration_response_payload_parsed.session_id)
+        };
+
+        // Poll both futures simultaneously to obtain the resulting socket and session_id
+        let (notary_socket, session_id) =
+            futures::try_join!(notary_connection_fut, client_requests_fut)?;
```

Comment:
@sinui0 was trying to use `futures::select!` so that i dont have to lump both http requests (and their req building, resp parsing) into one future, but to process the first `.send_request` future and `notary_connection_fut` independently — then I realised there's quite a bit of setup needed (e.g. Fuse and Pin required by select!), which might impact readability and maintenance, as I also can't see any upside of doing this v.s. `try_join!`,  but curious to know your thoughts.